### PR TITLE
fix: TX return data decoding for complex structs

### PR DESCRIPTION
**Pull request by ca11ab1e**
### What I did

Fix decoding complex structs.

### Checklist

- [x] Passes all linting checks (pre-commit and CI jobs)
- [x] New test cases have been added and are passing
- [ ] Documentation has been updated
- [x] PR title follows [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) standard (will be automatically included in the changelog)

### CODE REVIEW

**Comment by ca11ab1e** [comment]
I was thinking that a way to fix that is to be able to access the full contract ABI to get `ComplexStruct` details:
```js
StructABI(
    type='struct',
    name='ComplexStruct',
    members=[
        ABIType(name='timestamp', type='felt', components=None, internalType=None, offset=0),
        ABIType(name='value0', type='Uint256', components=None, internalType=None, offset=1),
        ABIType(name='value1', type='Uint256', components=None, internalType=None, offset=3)],
    size=5
)
```

But I did not find a way for now.

File: ape_starknet/provider.py [hunk 1]
```diff
@@ -277,7 +277,7 @@ def send_transaction(self, txn: TransactionAPI, token: Optional[str] = None) ->
                 abi = txn.original_method_abi
                 returndata = returndata[1:]
 
-            return_value = self.starknet.decode_returndata(abi, returndata)
+            return_value = self.starknet.decode_returndata(abi, returndata, [])
```
Comment (ca11ab1e) [inline ape_starknet/provider.py:280]: That's not good. We need the full aBI here too. It's the only missing piece, then we should be good @unparalleled-js.

Comment (unparalleled-js) [reply ape_starknet/provider.py]: Done  - the abi / contract-type is present on the txn

**Review by ca11ab1e — COMMENTED** [review]
tests: clean-up

File: ape_starknet/provider.py (hunk 1 above)
Comment (ca11ab1e) [reply ape_starknet/provider.py]: It's not enough. There is a full ABI, but that's the Account one, not the real one (ex: `full_abi = [StructABI(type='struct', name='AccountCallArray', ...]`). So structs are not present here.

I reminds me when the `txn.method_abi` is switched with the account-specific `execute_abi`. Should/Can we do the same for the full ABI?

Comment (ca11ab1e) [reply ape_starknet/provider.py]: Fixed with https://github.com/ApeWorX/ape-starknet/pull/69/commits/ddcde7ec2e67028bcf6115d66e0e7715cd535afa :champagne:

**Comment by ca11ab1e** [comment]
It's almost ready. It only needs to be synced with API change on the Ape side.

**Comment by ca11ab1e** [comment]
Last step: use the upcoming Ape release :)

File: ape_starknet/provider.py [hunk 2]
```diff
@@ -268,17 +268,18 @@ def send_transaction(self, txn: TransactionAPI, token: Optional[str] = None) ->
...
                 # We need to restore the former, and remove the later.
                 abi = txn.original_method_abi
                 returndata = returndata[1:]
-
-            return_value = self.starknet.decode_returndata(abi, returndata)
-            receipt.return_value = return_value
+                return_value = self.starknet.decode_returndata(
+                    abi, returndata, txn.original_full_abi
+                )
+                receipt.return_value = return_value
```
Comment (ca11ab1e) [inline ape_starknet/provider.py:282]: This is over indented :thinking: ?
```suggestion
            return_value = self.starknet.decode_returndata(
                abi, returndata, full_abi=txn.original_full_abi
            )
            receipt.return_value = return_value
```

File: ape_starknet/transactions.py [hunk 3]
```diff
@@ -149,14 +149,22 @@ def as_starknet_object(self) -> Deploy:
 class InvokeFunctionTransaction(StarknetTransaction):
     max_fee: int = 0
     method_abi: MethodABI
+
+    original_contract_type: Optional[ContractType] = None
```
Comment (ca11ab1e) [inline ape_starknet/transactions.py:153]: Where is it used?

File: ape_starknet/provider.py [hunk 4]
```diff
@@ -268,18 +268,20 @@ def send_transaction(self, txn: TransactionAPI, token: Optional[str] = None) ->
             receipt.returndata = returndata.copy()
             abi = txn.method_abi
 
-            if txn.original_method_abi and txn.original_full_abi:
+            if txn.original_method_abi:
                 # When those special attributes are set means the transaction came from an
                 # account-specific call: it implies original ABIs were replaced
                 # with account-specific ABIs in BaseStarknetAccount.prepare_transaction(),
                 # and that the return data is always prefixed with the number of items.
                 # We need to restore the former, and remove the later.
```
Comment (ca11ab1e) [inline ape_starknet/provider.py:276]:
```suggestion
                # When this special attribute is set means the transaction came from an
                # account-specific call: it implies original ABIs were replaced
                # with account-specific ABIs in BaseStarknetAccount.prepare_transaction(),
                # and that the return data is always prefixed with the number of items.
                # We need to restore the former, and remove the later.
```

File: ape_starknet/provider.py [hunk 5]
```diff
@@ -268,18 +268,20 @@ def send_transaction(self, txn: TransactionAPI, token: Optional[str] = None) ->
             receipt.returndata = returndata.copy()
             abi = txn.method_abi
```
Comment (ca11ab1e) [inline ape_starknet/provider.py:269]:
```suggestion

```

**Comment by ca11ab1e** [comment]
Can we also get ride of `txn.original_method_abi`?

**Comment by unparalleled-js** [comment]
> Can we also get ride of `txn.original_method_abi`?

I don't think we can

**Comment by ca11ab1e** [comment]
Approved :+1:

File: ape_starknet/ecosystems.py [hunk 6]
```diff
@@ -98,31 +97,14 @@ def decode_returndata(self, abi: MethodABI, raw_data: List[int]) -> Any:  # type
         if not raw_data:
             return raw_data
 
-        raw_data = [self.encode_primitive_value(v) for v in raw_data]
-        iter_data = iter(raw_data)
-        decoded: List[Any] = []
+        full_abi = [
+            a.dict() for a in (abi.contract_type.abi if abi.contract_type is not None else [abi])
+        ]
```
Comment (ca11ab1e) [inline ape_starknet/ecosystems.py:102]: `contract_abi` should always be available at that point:

```suggestion
        full_abi = [a.dict() for a in abi.contract_type.abi]  # type: ignore
```

Comment (unparalleled-js) [reply ape_starknet/ecosystems.py:102]: mypy does not know that

Comment (ca11ab1e) [reply ape_starknet/ecosystems.py:102]: That's the purpose of`type: ignore`.

**Review by unparalleled-js — APPROVED** [review]

**Pull request merged**